# Project 8 -- Lukas Anell

In [1]:
# /anvil/projects/tdm/data/spotify/taylor_swift_discography_updated.csv
# /anvil/projects/tdm/data/youtube/most_subscribed_youtube_channels.csv

## Question 1

In [2]:
import pandas as pd

In [3]:
ts_songs = pd.read_csv("/anvil/projects/tdm/data/spotify/taylor_swift_discography_updated.csv", sep=";")
ts_songs.head()

,ID,track_name,track_musical_genre,track_type,duration_ms,feature,track_videoclip,videoclip_views,spotify_streams,spotify_global_peak,...,uri,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
0,0,Fortnight (feat. Post Malone),Synth Pop,Single,228965,Post Malone,No,0,648525069,1,...,spotify:track:6dODwocEuGzHAavXqTbwHv,0.5020,0.504,0.386,1.53e-05,0.0961,-10.976,0.0308,192.004,0.281
1,1,The Tortured Poets Department,Synth Pop,B-Side,293048,No,No,0,206954130,6,...,spotify:track:4PdLaGZubp4lghChqp8erB,0.0483,0.604,0.428,0.0,0.1260,-8.441,0.0255,110.259,0.292
2,2,My Boy Only Breaks His Favorite Toys,Synth Pop,B-Side,203801,No,No,0,238016975,7,...,spotify:track:7uGYWMwRy24dm7RUDDhUlD,0.1370,0.596,0.563,0.0,0.3020,-7.362,0.0269,97.073,0.481
3,3,Down Bad,R&B,B-Side,261228,No,No,0,318375377,2,...,spotify:track:1kbEbBdEgQdQeLXCJh28pJ,0.5600,0.541,0.366,"1,00E-06",0.0946,-10.412,0.0748,159.707,0.168
4,4,"So Long, London",Synth Pop,B-Side,262974,No,No,0,253712712,4,...,spotify:track:7wAkQFShJ27V8362MqevQr,0.7300,0.423,0.533,0.00264,0.0816,-11.388,0.3220,160.218,0.248


In [24]:
ts_songs.shape

(577, 28)

In [8]:
def find_songs_with_energy(input_df, threshold):
    my_output = input_df[input_df["energy"] >= threshold]
    return my_output

In [15]:
my_median = ts_songs['energy'].median()
my_median

0.57

In [21]:
ts_songs['energy'].max()

0.95

In [20]:
ts_songs['energy'].min()

0.118

In [22]:
high_energy_df = find_songs_with_energy(ts_songs, my_median)
high_energy_df.shape

(289, 28)

In [23]:
high_energy_df.head()

,ID,track_name,track_musical_genre,track_type,duration_ms,feature,track_videoclip,videoclip_views,spotify_streams,spotify_global_peak,...,uri,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
5,5,But Daddy I Love Him,Pop Rock,B-Side,340428,No,No,0,229410536,9,...,spotify:track:4QMgEffJQuKtjCNvqfRZ0m,0.3840,0.521,0.720,0.0,0.135,-7.684,0.1040,79.943,0.438
7,7,Florida!!! (feat. Florence + The Machine),Indie Rock,B-Side,215463,Florence + The Machine,No,0,201272168,10,...,spotify:track:5ExOm0dh4NyRyAdSAO9hyM,0.1780,0.533,0.573,0.0,0.309,-7.123,0.1380,150.088,0.398
12,12,I Can Do It With a Broken Heart,Electropop,Single,218004,No,No,0,439965329,5,...,spotify:track:2fPvQfGQEZOKtJ9qXeL4x8,0.1230,0.701,0.751,0.0,0.150,-8.390,0.0502,129.994,0.220
21,21,So High School,Pop Rock,B-Side,228800,No,No,0,169310310,29,...,spotify:track:7Mts0OfPorF4iwOomvfqn1,0.0274,0.366,0.866,3.07e-06,0.106,-4.514,0.0466,176.070,0.293
26,26,Cassandra,Folk Pop,B-Side,240205,No,No,0,77285946,54,...,spotify:track:0g4fMVo4JjwnIpTfFfLdxS,0.8060,0.587,0.587,0.0,0.137,-9.879,0.0595,139.963,0.514


In [4]:
def find_songs_by_album(df, album_name):
    my_output = df[df["album"] == album_name]
    return my_output

In [5]:
speak_now = find_songs_by_album(ts_songs, "Speak Now")
speak_now.shape

(14, 28)

In [6]:
speak_now.head()

,ID,track_name,track_musical_genre,track_type,duration_ms,feature,track_videoclip,videoclip_views,spotify_streams,spotify_global_peak,...,uri,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
508,508,Mine,Country Pop,Single,230706,No,Yes,319.358.547,220.866.384,0,...,spotify:track:0dBW6ZsW8skfvoRfgeerBF,0.00265,0.624,0.757,1.87e-06,0.1890,-2.940,0.0296,121.070,0.658
509,509,Sparks Fly,Country Pop,Single,260933,No,Yes,97.248.044,180.160.266,0,...,spotify:track:6d9IiDcFxtFVIvt9pCqyGH,0.03960,0.605,0.787,1.42e-06,0.1630,-3.002,0.0308,114.987,0.374
510,510,Back To December,Country Pop,Single,293026,No,Yes,328.789.172,386.437.949,0,...,spotify:track:3DrjZArsPsoqbLzUZZV1Id,0.11700,0.529,0.670,0.0,0.3340,-4.663,0.0303,141.893,0.286
511,511,Speak Now,Country Pop,B-Side,240760,No,No,0,142.918.741,0,...,spotify:track:24DefNCFiWTP8OjYWiXuYe,0.09500,0.709,0.599,0.0,0.0973,-3.734,0.0304,118.975,0.735
512,512,Dear John,Country Pop,B-Side,403920,No,No,0,140.397.028,0,...,spotify:track:7hZuICN5eaCuQyp443RCt6,0.16600,0.589,0.470,1.45e-06,0.1120,-5.320,0.0280,119.386,0.102


In [13]:
example = find_songs_by_album(ts_songs, "The Tortured Poets Department: The Anthology")
example.shape

(31, 28)

In [14]:
example.head()

,ID,track_name,track_musical_genre,track_type,duration_ms,feature,track_videoclip,videoclip_views,spotify_streams,spotify_global_peak,...,uri,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
0,0,Fortnight (feat. Post Malone),Synth Pop,Single,228965,Post Malone,No,0,648525069,1,...,spotify:track:6dODwocEuGzHAavXqTbwHv,0.5020,0.504,0.386,1.53e-05,0.0961,-10.976,0.0308,192.004,0.281
1,1,The Tortured Poets Department,Synth Pop,B-Side,293048,No,No,0,206954130,6,...,spotify:track:4PdLaGZubp4lghChqp8erB,0.0483,0.604,0.428,0.0,0.1260,-8.441,0.0255,110.259,0.292
2,2,My Boy Only Breaks His Favorite Toys,Synth Pop,B-Side,203801,No,No,0,238016975,7,...,spotify:track:7uGYWMwRy24dm7RUDDhUlD,0.1370,0.596,0.563,0.0,0.3020,-7.362,0.0269,97.073,0.481
3,3,Down Bad,R&B,B-Side,261228,No,No,0,318375377,2,...,spotify:track:1kbEbBdEgQdQeLXCJh28pJ,0.5600,0.541,0.366,"1,00E-06",0.0946,-10.412,0.0748,159.707,0.168
4,4,"So Long, London",Synth Pop,B-Side,262974,No,No,0,253712712,4,...,spotify:track:7wAkQFShJ27V8362MqevQr,0.7300,0.423,0.533,0.00264,0.0816,-11.388,0.3220,160.218,0.248


1.1: The max energy was 0.95, and the min was 0.118.

1.2: 289 songs had high energy.

1.3: See above.

## Question 2

In [26]:
# song durations in seconds
ts_songs['duration_sec'] = ts_songs['duration_ms'] / 1000
# song durations in minutes
ts_songs['duration_min'] = ts_songs['duration_sec'] / 60

In [29]:
total_by_album = ts_songs.groupby("album")["duration_min"].sum().reset_index()
total_by_album.head()

,album,duration_min
0,1989,48.797733
1,1989 (Deluxe Edition),68.760800
2,1989 (Taylor's Version),77.972117
3,1989 (Taylor's Version) [Deluxe],81.301667
4,Fearless,53.547900


In [31]:
long_albums = total_by_album[total_by_album["duration_min"] > 80]
long_albums.head()

,album,duration_min
3,1989 (Taylor's Version) [Deluxe],81.301667
5,Fearless (Taylor's Version),106.541500
11,Midnights (The Til Dawn Edition),80.586083
13,Red (Deluxe Edition),90.217667
14,Red (Taylor's Version),130.663833


In [32]:
def find_albums_by_time(input_df, input_min_duration):
    total_by_album = input_df.groupby('album')['duration_min'].sum() 
    long_albums = total_by_album[total_by_album > input_min_duration] 

    return long_albums

In [33]:
min_80 = find_albums_by_time(ts_songs, 80)
min_80.head()

album
1989 (Taylor's Version) [Deluxe]     81.301667
Fearless (Taylor's Version)         106.541500
Midnights (The Til Dawn Edition)     80.586083
Red (Deluxe Edition)                 90.217667
Red (Taylor's Version)              130.663833
Name: duration_min, dtype: float64

In [34]:
min_120 = find_albums_by_time(ts_songs, 120)
min_120.head()

album
Red (Taylor's Version)                                                                 130.663833
The Tortured Poets Department: The Anthology                                           122.645733
folklore: the long pond studio sessions (from the Disney+ special) [deluxe edition]    134.704517
reputation Stadium Tour Surprise Song Playlist                                         186.266467
Name: duration_min, dtype: float64

2.1: See above.

2.2: See above.

2.3: See above.

## Question 3

In [35]:
lower_threshold = 30
upper_threshold = 80

In [37]:
def classify(duration):
        if duration < lower_threshold:
            return "short_name"
        elif duration > upper_threshold:
            return "long_name"
        else:
            return "other/default_name"

In [39]:
total_by_album = ts_songs.groupby("album")["duration_min"].sum().reset_index()   

In [41]:
total_by_album["duration_min"].apply(classify).head()

0    other/default_name
1    other/default_name
2    other/default_name
3             long_name
4    other/default_name
Name: duration_min, dtype: object

In [36]:
def albums_by_length(df, lower_threshold, upper_threshold):
    total_by_album = df.groupby("album")["duration_min"].sum().reset_index()                
    
    def classify(duration):
        if duration < lower_threshold:
            return "short_name"
        elif duration > upper_threshold:
            return "long_name"
        else:
            return "other/default_name"
    
    total_by_album["category"] = total_by_album["duration_min"].apply(classify)
    
    return total_by_album

In [42]:
test = albums_by_length(ts_songs, lower_threshold=30, upper_threshold=80)
test.head()

,album,duration_min,category
0,1989,48.797733,other/default_name
1,1989 (Deluxe Edition),68.760800,other/default_name
2,1989 (Taylor's Version),77.972117,other/default_name
3,1989 (Taylor's Version) [Deluxe],81.301667,long_name
4,Fearless,53.547900,other/default_name


3.1: See above.

3.2: See above.

3.3: See above.

## Question 4

In [43]:
youtubers = pd.read_csv("/anvil/projects/tdm/data/youtube/most_subscribed_youtube_channels.csv")
youtubers.head()

,rank,Youtuber,subscribers,video views,video count,category,started
0,1,T-Series,"222,000,000","198,459,090,822","17,317",Music,2006
1,2,YouTube Movies,"154,000,000",0,0,Film & Animation,2015
2,3,Cocomelon - Nursery Rhymes,"140,000,000","135,481,339,848",786,Education,2006
3,4,SET India,"139,000,000","125,764,252,686","91,271",Shows,2006
4,5,Music,"116,000,000",0,0,NaN,2013


In [45]:
youtubers["video_count2"] = pd.to_numeric(youtubers["video count"].str.replace(",", "", regex=False))
youtubers.head()

,rank,Youtuber,subscribers,video views,video count,category,started,video_count2
0,1,T-Series,"222,000,000","198,459,090,822","17,317",Music,2006,17317
1,2,YouTube Movies,"154,000,000",0,0,Film & Animation,2015,0
2,3,Cocomelon - Nursery Rhymes,"140,000,000","135,481,339,848",786,Education,2006,786
3,4,SET India,"139,000,000","125,764,252,686","91,271",Shows,2006,91271
4,5,Music,"116,000,000",0,0,NaN,2013,0


In [48]:
youtubers.groupby('category')['video_count2'].sum().sort_values(ascending=False)

category
News & Politics          2754693
Entertainment            2674176
People & Blogs           1045091
Music                     510337
Gaming                    427292
Shows                     283027
Nonprofits & Activism     188445
Sports                    140464
Film & Animation          133319
Education                 124727
Comedy                     93562
Howto & Style              81419
Science & Technology       36622
Pets & Animals             23960
Trailers                   13613
Movies                      5576
Autos & Vehicles            2874
Travel & Events              632
Name: video_count2, dtype: int64

In [46]:
gm_subset = youtubers[youtubers["category"].isin(["Gaming", "Music"])]
gm_subset.head()

,rank,Youtuber,subscribers,video views,video count,category,started,video_count2
0,1,T-Series,"222,000,000","198,459,090,822","17,317",Music,2006,17317
5,6,PewDiePie,"111,000,000","28,469,458,228","4,497",Gaming,2010,4497
11,12,Zee Music Company,"86,700,000","49,067,711,243","6,603",Music,2014,6603
18,19,BANGTANTV,"70,000,000","17,471,090,750","1,787",Music,2012,1787
19,20,Justin Bieber,"69,700,000","28,567,240,999",247,Music,2007,247


In [47]:
grouped_counts = gm_subset.groupby(["category", "started"])["video_count2"].sum().unstack()
grouped_counts.head()

started,2005,2006,2007,2008,2009,2010,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020
category,,,,,,,,,,,,,,,,
Gaming,17878.0,163116.0,11967.0,16356.0,15342.0,9998.0,27421.0,72858.0,32267.0,31332.0,7342.0,7903.0,9302.0,2952.0,1126.0,132.0
Music,1446.0,36423.0,51621.0,33451.0,16392.0,31766.0,89353.0,44299.0,71681.0,119692.0,5447.0,8477.0,156.0,121.0,12.0,NaN


4.1: See above.

4.2: See above.

4.3: See above.

## Question 5

In [49]:
youtubers['subscribers2'] = youtubers['subscribers'].str.replace(',', '').astype(float)
youtubers.head()

,rank,Youtuber,subscribers,video views,video count,category,started,video_count2,subscribers2
0,1,T-Series,"222,000,000","198,459,090,822","17,317",Music,2006,17317,222000000.0
1,2,YouTube Movies,"154,000,000",0,0,Film & Animation,2015,0,154000000.0
2,3,Cocomelon - Nursery Rhymes,"140,000,000","135,481,339,848",786,Education,2006,786,140000000.0
3,4,SET India,"139,000,000","125,764,252,686","91,271",Shows,2006,91271,139000000.0
4,5,Music,"116,000,000",0,0,NaN,2013,0,116000000.0


In [50]:
def top_youtuber_by_genre(df, genre):
    genre_rows = df[df['category'] == genre]
    top_index = genre_rows['subscribers2'].idxmax()

    return genre_rows.loc[top_index]

In [51]:
top_gaming = top_youtuber_by_genre(youtubers, "Gaming")
print(top_gaming)

rank                         6
Youtuber             PewDiePie
subscribers        111,000,000
video views     28,469,458,228
video count              4,497
category                Gaming
started                   2010
video_count2              4497
subscribers2       111000000.0
Name: 5, dtype: object


In [52]:
top_gaming = top_youtuber_by_genre(youtubers, "Music")
print(top_gaming)

rank                          1
Youtuber               T-Series
subscribers         222,000,000
video views     198,459,090,822
video count              17,317
category                  Music
started                    2006
video_count2              17317
subscribers2        222000000.0
Name: 0, dtype: object


5.1: See above.

5.2: The second category I used was Music, and T-Series was returned.

5.3: I'm not up to date with the current list, but I think it's definitely changed a lot with youtubers like Mr. Beast or other large channels that have come up since this dataset.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
